### PR TITLE
Fortinet Firewall Object - Virtual IPs

### PR DESCRIPTION
**Pull request by ftntcorecse**
##### SUMMARY

Allows the add/set/delete of Virtual IP Objects within FortiManager

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

fmgr_fwobj_vip.py
test_fmgr_fwobj_vip.py
fixtures/test_fmgr_fwobj_vip.json
##### ANSIBLE VERSION

```paste below
ansible 2.8.0.dev0 (devel 9efc3dc761) last updated 2018/09/27 20:21:00 (GMT -700)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/usr/ansible_ftnt/library']
  ansible python module location = /root/ansibleSource/ansible/lib/ansible
  executable location = /root/ansibleSource/ansible/bin/ansible
  python version = 2.6.9 (unknown, Sep 27 2018, 21:07:46) [GCC 5.4.0 20160609]
```

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

File: lib/ansible/modules/network/fortimanager/fmgr_fwobj_vip.py [hunk 1]
```diff
@@ -0,0 +1,2595 @@
...
+    required: false
+    choices: ["ssl-3.0", "tls-1.0", "tls-1.1", "tls-1.2", "client"]
+
+  ssl_server_algorithm:
+    description:
+      - Permitted encryption algorithms for the server side of SSL full mode sessions according to encryption strength
+      - choice | high | High encryption. Allow only AES and ChaCha.
+      - choice | low | Low encryption. Allow AES, ChaCha, 3DES, RC4, and DES.
+      - choice | medium | Medium encryption. Allow AES, ChaCha, 3DES, and RC4.
+      - choice | custom | Custom encryption. Use ssl-server-cipher-suites to select the cipher suites that are allow
```
Comment (samccann) [inline lib/ansible/modules/network/fortimanager/fmgr_fwobj_vip.py]:
```suggestion
      - choice | custom | Custom encryption. Use ssl-server-cipher-suites to select the cipher suites that are allowed.
```

File: lib/ansible/modules/network/fortimanager/fmgr_fwobj_vip.py [hunk 2]
```diff
@@ -0,0 +1,2595 @@
...
+      - choice | client | Use the same encryption algorithms for both client and server sessions.
+    required: false
+    choices: ["high", "low", "medium", "custom", "client"]
+
+  ssl_send_empty_frags:
+    description:
+      - Enable/disable sending empty fragments to avoid CBC IV attacks (SSL 3.0 &amp; TLS 1.0 only). May need to be di
+      - sabled for compatibility with older systems.
```
Comment (samccann) [inline lib/ansible/modules/network/fortimanager/fmgr_fwobj_vip.py]: There are a couple like this where the word is split between two lines. I don't know if this renders correctly or not when the docs are build? Might be better to breake line 151 at the period and move the whole 2nd sentence to line 152.

File: lib/ansible/modules/network/fortimanager/fmgr_fwobj_vip.py [hunk 3]
```diff
@@ -0,0 +1,2595 @@
...
+      - choice | disable | Do not send empty fragments.
+      - choice | enable | Send empty fragments.
+    required: false
+    choices: ["disable", "enable"]
+
+  ssl_pfs:
+    description:
+      - Select the cipher suites that can be used for SSL perfect forward secrecy (PFS). Applies to both client and se
+      - rver sessions.
```
Comment (samccann) [inline lib/ansible/modules/network/fortimanager/fmgr_fwobj_vip.py]: same issue here with a truncated word.

File: lib/ansible/modules/network/fortimanager/fmgr_fwobj_vip.py [hunk 4]
```diff
@@ -0,0 +1,2595 @@
...
+    description:
+      - Maximum number of client to FortiGate SSL session states to keep.
+    required: false
+
+  ssl_client_renegotiation:
+    description:
+      - Allow, deny, or require secure renegotiation of client sessions to comply with RFC 5746.
+      - choice | deny | Abort any client initiated SSL re-negotiation attempt.
+      - choice | allow | Allow a SSL client to renegotiate.
+      - choice | secure | Abort any client initiated SSL re-negotiation attempt that does not use RFC 5746 Secure Re
```
Comment (samccann) [inline lib/ansible/modules/network/fortimanager/fmgr_fwobj_vip.py]: Truncated line.. not sure what Secure Re should say...

File: lib/ansible/modules/network/fortimanager/fmgr_fwobj_vip.py [hunk 5]
```diff
@@ -0,0 +1,2595 @@
...
+
+  srcintf_filter:
+    description:
+      - Interfaces to which the VIP applies. Separate the names with spaces.
+    required: false
+
+  src_filter:
+    description:
+      - Source address filter. Each address must be either an IP/subnet (x.x.x.x/n) or a range (x.x.x.x-y.y.y.y). Sepa
+      - rate addresses with spaces.
```
Comment (samccann) [inline lib/ansible/modules/network/fortimanager/fmgr_fwobj_vip.py]: truncated word should be moved down to line 343.

File: lib/ansible/modules/network/fortimanager/fmgr_fwobj_vip.py [hunk 6]
```diff
@@ -0,0 +1,2595 @@
...
+      - Enable to add the Front-End-Https header for Microsoft Outlook Web Access.
+      - choice | disable | Disable Outlook Web Access support.
+      - choice | enable | Enable Outlook Web Access support.
+    required: false
+    choices: ["disable", "enable"]
+
+  nat_source_vip:
+    description:
+      - Enable to prevent unintended servers from using a virtual IP. Disable to use the actual IP address of the serv
+      - er as the source address.
```
Comment (samccann) [inline lib/ansible/modules/network/fortimanager/fmgr_fwobj_vip.py]: truncated word should be moved down to line 413.

File: lib/ansible/modules/network/fortimanager/fmgr_fwobj_vip.py [hunk 7]
```diff
@@ -0,0 +1,2595 @@
...
+      - Enable/disable HTTP multiplexing.
+      - choice | disable | Disable HTTP session multiplexing.
+      - choice | enable | Enable HTTP session multiplexing.
+    required: false
+    choices: ["disable", "enable"]
+
+  http_ip_header_name:
+    description:
+      - For HTTP multiplexing, enter a custom HTTPS header name. The original client IP address is added to this heade
+      - r. If empty, X-Forwarded-For is used.
```
Comment (samccann) [inline lib/ansible/modules/network/fortimanager/fmgr_fwobj_vip.py]: truncated word should be moved down to line 481

File: lib/ansible/modules/network/fortimanager/fmgr_fwobj_vip.py [hunk 8]
```diff
@@ -0,0 +1,2595 @@
...
+
+  extport:
+    description:
+      - Incoming port number range that you want to map to a port number range on the destination network.
+    required: false
+
+  extip:
+    description:
+      - IP address or address range on the external interface that you want to map to an address or address range on t
+      - he destination network.
```
Comment (samccann) [inline lib/ansible/modules/network/fortimanager/fmgr_fwobj_vip.py:539]: truncated word should be moved to line 542

File: lib/ansible/modules/network/fortimanager/fmgr_fwobj_vip.py [hunk 9]
```diff
@@ -0,0 +1,2595 @@
...
+      - choice | disable | Disable per server health check.
+      - choice | enable | Enable per server health check.
+      - choice | vip | Use health check defined in VIP.
+    required: false
+    choices: ["disable", "enable", "vip"]
+
+  realservers_holddown_interval:
+    description:
+      - Time in seconds that the health check monitor continues to monitor and unresponsive server that should be acti
+      - ve.
```
Comment (samccann) [inline lib/ansible/modules/network/fortimanager/fmgr_fwobj_vip.py]: truncated word.

File: lib/ansible/modules/network/fortimanager/fmgr_fwobj_vip.py [hunk 10]
```diff
@@ -0,0 +1,2595 @@
...
+
+  realservers_ip:
+    description:
+      - IP address of the real server.
+    required: false
+
+  realservers_max_connections:
+    description:
+      - Max number of active connections that can be directed to the real server. When reached, sessions are sent to o
+      - ther real servers.
```
Comment (samccann) [inline lib/ansible/modules/network/fortimanager/fmgr_fwobj_vip.py]: truncated word

File: lib/ansible/modules/network/fortimanager/fmgr_fwobj_vip.py [hunk 11]
```diff
@@ -0,0 +1,2595 @@
...
+
+  realservers_seq:
+    description:
+      - Real Server Sequence Number
+    required: false
+
+  realservers_status:
+    description:
+      - Set the status of the real server to active so that it can accept traffic, or on standby or disabled so no tra
+      - ffic is sent.
```
Comment (samccann) [inline lib/ansible/modules/network/fortimanager/fmgr_fwobj_vip.py]: truncated word

File: lib/ansible/modules/network/fortimanager/fmgr_fwobj_vip.py (hunk 2 above)
Comment (ftntcorecse) [reply lib/ansible/modules/network/fortimanager/fmgr_fwobj_vip.py]: Nail on the head. This is a rendering issue from our API docs. Will fix in our workflow this week. For now, spot checking.

**Review by samccann — APPROVED** [review]
docs are good.

File: lib/ansible/modules/network/fortimanager/fmgr_fwobj_vip.py [hunk 12]
```diff
@@ -0,0 +1,2259 @@
...
+      - Enable/disable sending empty fragments to avoid CBC IV attacks (SSL 3.0 &amp; TLS 1.0 only).
+      - sabled for compatibility with older systems.
+      - choice | disable | Do not send empty fragments.
+      - choice | enable | Send empty fragments.
+    required: false
+    choices: ["disable", "enable"]
+
+  ssl_pfs:
+    description:
+      - Select the cipher suites that can be used for SSL perfect forward secrecy (PFS). Applies to both client and
```
Comment (NilashishC) [inline lib/ansible/modules/network/fortimanager/fmgr_fwobj_vip.py]: This line seems to have been split into two.

File: lib/ansible/modules/network/fortimanager/fmgr_fwobj_vip.py [hunk 13]
```diff
@@ -0,0 +1,2259 @@
...
+      - server sessions.
+      - choice | require | Allow only Diffie-Hellman cipher-suites, so PFS is applied.
+      - choice | deny | Allow only non-Diffie-Hellman cipher-suites, so PFS is not applied.
+      - choice | allow | Allow use of any cipher suite so PFS may or may not be used depending on the cipher suite.
+    required: false
+    choices: ["require", "deny", "allow"]
+
+  ssl_mode:
+    description:
+      - Apply SSL offloading between the client and the FortiGate (half) or from the client to the FortiGate and from
```
Comment (NilashishC) [inline lib/ansible/modules/network/fortimanager/fmgr_fwobj_vip.py]: Looks like a split line.

File: lib/ansible/modules/network/fortimanager/fmgr_fwobj_vip.py [hunk 14]
```diff
@@ -0,0 +1,2259 @@
...
+
+
+###############
+# START METHODS
+###############
+
+
+def fmgr_firewall_vip_addsetdelete(fmg, paramgram):
+    """
+    fmgr_firewall_vip -- Your Description here, bruh
```
Comment (NilashishC) [inline lib/ansible/modules/network/fortimanager/fmgr_fwobj_vip.py]: Please modify the generic function description statement.

**Review by NilashishC — APPROVED** [review]

**Comment by NilashishC** [comment]
LGTM

**Pull request merged**